### PR TITLE
ffmpeg: general option improvements

### PR DESCRIPTION
**Pull request by jopejoe1**
## Description of changes

- add options for some build features (static, shared, pic, thumb)
- move libdc1394 and libraw1394 features flag (only needed when dc1394 enabled)
- remove gplv3 option (this is the case if version3 and gpl are both activated)
- add version3 option (doesn't need to be gpl3 lpgl3 also can be the case)
- added the right license requirements to options form (https://git.ffmpeg.org/gitweb/ffmpeg.git/blob/8e23ebe6f971ec4177be73b453960d83c264b7b5:/configure#l1804)

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/development/libraries/ffmpeg/generic.nix [hunk 1]
```diff
@@ -302,11 +302,6 @@ in
 
 assert lib.elem ffmpegVariant [ "headless" "small" "full" ];
 
-/*
- *  Licensing dependencies
- */
-assert withGPLv3 -> withGPL;
-assert withUnfree -> withGPL && withGPLv3;
```
Comment (Atemu) [inline pkgs/development/libraries/ffmpeg/generic.nix:304]: Oh, this was actually wrong. It should probably be re-instated as

```Nix
assert !(withUnfree && !withGPL);
```

as these are mutually exclusive.

File: pkgs/development/libraries/ffmpeg/generic.nix [hunk 2]
```diff
@@ -71,7 +71,7 @@
 , withPulse ? withSmallDeps && !stdenv.isDarwin # Pulseaudio input support
 , withRav1e ? withFullDeps # AV1 encoder (focused on speed and safety)
 , withRtmp ? false # RTMP[E] support
-, withSamba ? withFullDeps && !stdenv.isDarwin # Samba protocol
+, withSamba ? withFullDeps && !stdenv.isDarwin && withGPL && withVersion3 # Samba protocol
```
Comment (Atemu) [inline pkgs/development/libraries/ffmpeg/generic.nix:74]: I think I might prefer a `withGPLv3` alias for this.

File: pkgs/development/libraries/ffmpeg/generic.nix [hunk 3]
```diff
@@ -85,23 +85,23 @@
 , withV4l2M2m ? withV4l2
 , withVaapi ? withHeadlessDeps && (with stdenv; isLinux || isFreeBSD) # Vaapi hardware acceleration
 , withVdpau ? withSmallDeps # Vdpau hardware acceleration
-, withVidStab ? withFullDeps # Video stabilization
-, withVmaf ? withFullDeps && withGPLv3 && !stdenv.isAarch64 # Netflix's VMAF (Video Multi-Method Assessment Fusion)
-, withVoAmrwbenc ? withFullDeps # AMR-WB encoder
+, withVidStab ? withFullDeps && withGPL # Video stabilization
```
Comment (Atemu) [inline pkgs/development/libraries/ffmpeg/generic.nix:88]: Vidstab is MIT AFAICT?

File: pkgs/development/libraries/ffmpeg/generic.nix [hunk 4]
```diff
@@ -85,23 +85,23 @@
...
+, withVmaf ? withFullDeps && !stdenv.isAarch64 # Netflix's VMAF (Video Multi-Method Assessment Fusion)
```
Comment (Atemu) [inline pkgs/development/libraries/ffmpeg/generic.nix:89]: (This will conflict soon.)

**Review by Atemu — COMMENTED** [review]
I was going to NAK because putting two licenses behind a single flag like that is poor design but it's actually upstream who did it like that.

File: pkgs/development/libraries/ffmpeg/generic.nix (hunk 3 above)
Comment (jopejoe1) [reply pkgs/development/libraries/ffmpeg/generic.nix:88]: ffmpeg has it under gpl and looking at theire source it also says gpl https://github.com/georgmartius/vid.stab/blob/master/LICENSE

File: pkgs/development/libraries/ffmpeg/generic.nix (hunk 4 above)
Comment (jopejoe1) [reply pkgs/development/libraries/ffmpeg/generic.nix:89]: Noted will rebase when that gets merged

File: pkgs/development/libraries/ffmpeg/generic.nix (hunk 1 above)
Comment (jopejoe1) [reply pkgs/development/libraries/ffmpeg/generic.nix:304]: You can compile ffmpeg with both of them enabled, but the resulting binary becomes unredistributable, gone updated the license list to correctly reflect that

File: pkgs/development/libraries/ffmpeg/generic.nix (hunk 4 above)
Comment (jopejoe1) [reply pkgs/development/libraries/ffmpeg/generic.nix]: rebased

File: pkgs/development/libraries/ffmpeg/generic.nix (hunk 3 above)
Comment (Atemu) [reply pkgs/development/libraries/ffmpeg/generic.nix:88]: Ah, I was looking at the python module with the same name.

File: pkgs/development/libraries/ffmpeg/generic.nix [hunk 5]
```diff
@@ -376,14 +382,15 @@ stdenv.mkDerivation (finalAttrs: {
...
     /*
      *  Build flags
      */
-    # On some ARM platforms --enable-thumb
-    "--enable-shared"
-    "--enable-pic"
+    (enableFeature withStatic "static")
+    (enableFeature withShared "shared")
+    (enableFeature withPic "pic")
+    (enableFeature withThumb "thumb")
```
Comment (Atemu) [inline pkgs/development/libraries/ffmpeg/generic.nix:393]: Have you verified that these work?

**Review by Atemu — APPROVED** [review]
Looks pretty good to me now.

File: pkgs/development/libraries/ffmpeg/generic.nix (hunk 5 above)
Comment (jopejoe1) [reply pkgs/development/libraries/ffmpeg/generic.nix:393]: Tested static and shared (works if one or both are enabled, static only tested on headless), chaining pic works with static on shared it needs to enabled other wise the build fails, can not test enabling static as i do not own a 32-bit arm machine with thumb support

**Review by arthsmn — APPROVED** [review]

File: pkgs/development/libraries/ffmpeg/generic.nix [hunk 6]
```diff
@@ -703,8 +708,10 @@ stdenv.mkDerivation (finalAttrs: {
     '';
     license = with licenses; [ lgpl21Plus ]
```
Comment (AndersonTorres) [inline pkgs/development/libraries/ffmpeg/generic.nix:709]: I don't know how license works (IANAL), but it looks a bit self-contradictory to include free licenses with unfree ones.
It should be something like "if unfree components were selected then license is unfree; else, insert free licenses accordingly".

Comment (jopejoe1) [reply pkgs/development/libraries/ffmpeg/generic.nix:709]: I think its still useful to know if there is lgpl/gpl code included in a binary even if the resulting binary is unfree (which is not even true in all cases where the unfree option is enabled)

Comment (Atemu) [reply pkgs/development/libraries/ffmpeg/generic.nix:709]: The free licenses also all still apply to at least some parts of the code. Distributors must accompany the artifacts with the license text and copyright notices with most FOSS licenses. IANAL but I think we're fulfilling our legal obligations here.

The real "license" of a package is also always the composite of all of these licenses. If you don't accept one of them, you don't even have the right to download and run the package.  

This nature of licenses is also important in cases such as i.e. `[ unfreeRedistributable gpl3Plus ]` where you *cannot* actually redistribute the resulting package as the GPLv3 only allows redistribution according to its terms for the entire code-base which are not compatible with the terms of unfreeRedistributable.

We need a proper system for these composite license due to problems like this. I'll open an issue on that.

**Review by arthsmn — APPROVED** [review]

**Review by tilpner — APPROVED** [review]

File: pkgs/development/libraries/ffmpeg/generic.nix (hunk 6 above)
Comment (Atemu) [reply pkgs/development/libraries/ffmpeg/generic.nix:709]: Forgot to add a back link: https://github.com/NixOS/nixpkgs/issues/283706

**Comment by Atemu** [comment]
Thanks :)

**Pull request merged**